### PR TITLE
refactor(ica): packet data unmarshaling logic refactored

### PR DESCRIPTION
**Pull request by srdtrk**
## Description

This PR refactors the packet data unmarshaling logic. Once merged, this pr can be used as an example when refactoring other apps similarly.

closes: #4224 

### Commit Message / Changelog Entry

```bash
refactor(ica): packet data unmarshaling logic refactored
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/controller/ibc_middleware.go:264]: We could also make `UnmarshalJSON` a method of `InterchainAccountPacketData`. 

```go
func (iapd InterchainAccountPacketData) UnmarshalJSON(bz []byte) (InterchainAccountPacketData, error)
```

Usage:

```go
var packetData icatypes.InterchainAccountPacketData
if err := packetData.UnmarshalJSON(bz); err != nil { 
   // handle err
}
```

Happy to defer to you guys on what API you prefer

**Review by damiannolan — COMMENTED** [review]
All looks good, I left a suggestion on an alternative API. Happy to let you guys go with whatever you prefer 👍

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:264]: Yes, this is also a valid approach although the function signature likely should be different. I tried this first (and called it `FromBytes`) but I thought the one I added looked better in the end. I'm happy to keep this PR open until we get some other opinions

```go
func (iapd *InterchainAccountPacketData) FromBytes(bz []byte) error
```

Comment (damiannolan) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:264]: > although the function signature likely should be different

Aha, yes, sorry for my typo! Indeed, you don't need to return `(InterchainAccountPacketData, error)` since you're acting on unmarshaling into the receiver. I'll edit my first comment! 😛 

I think `UnmarshalJSON` naming is pretty common/standard in golang and in cosmos. Here's an example in [`x/staking`](https://github.com/cosmos/cosmos-sdk/blob/60198f077af73867594d8dbe0359e5f84b8cba2d/x/staking/types/delegation.go#L312) and [`x/auth`](https://github.com/cosmos/cosmos-sdk/blob/60198f077af73867594d8dbe0359e5f84b8cba2d/x/auth/types/account.go#L259)

Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:264]: Ok, sounds good. I'll do this

**Review by DimitrisJim — APPROVED** [review]
way cleaner :100:

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go]: Note: I did this

**Review by chatton — APPROVED** [review]
LGTM nice improvement 👍

File: modules/apps/27-interchain-accounts/types/packet.go
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/types/packet.go:60]: I am fine with this approach, but maybe an alternative could be to have instead and constructor function if that's possible? Like this:

```go
func NewInterchainAccountPacketData(bz []byte) (InterchainAccountPacketData, error) {
  var packetData icatypes.InterchainAccountPacketData
  if err := ModuleCdc.UnmarshalJSON(bz, &packetData); err != nil {
    return nil, err
  }
  return packetData, nil
}
```

It just feels more natural to me that declaring the value and then mutating it. That way we don't mix value and pointer receiver functions in the type, which I think it's another go proverb? :)

Comment (srdtrk) [reply modules/apps/27-interchain-accounts/types/packet.go:60]: This is exactly what I did originally, but @damiannolan and @DimitrisJim were slightly favouring the current approach. I'm curious what others think. Regardless, I apparently can't merge this until ica code owners approve it

Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/types/packet.go:60]: Ah, funny, haha! :) Well, if the majority prefers the current approach, I am ok with it.

Comment (damiannolan) [inline modules/apps/27-interchain-accounts/types/packet.go:60]: Could we add a godoc? :)

**Review by damiannolan — APPROVED** [review]
Thanks @srdtrk!

File: modules/apps/27-interchain-accounts/types/packet.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/types/packet.go:61]: done

**Pull request merged**
